# Chapter 4, inference / hypothesis testing

## C1

In [ ]:
import wooldridge
import pandas as pd
import numpy as np
from statsmodels.formula.api import ols
import matplotlib.pyplot as plt


# Cargar datos
data = wooldridge.data('vote1')

vote = pd.DataFrame(data)


In [16]:
# Crear el modelo 
model = ols('voteA ~ lexpendA + lexpendB + prtystrA', data=vote).fit()

In [18]:
# Print model summary
print (model.summary())

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Mon, 11 Nov 2024   Prob (F-statistic):           1.76e-57
Time:                        17:27:40   Log-Likelihood:                -596.86
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     45.0789      3.926     11.481      0.0

(i) **interpretation of b1**
b1 models a level-log relationship, but vote is a percentage, a change of 1% in expendA changes the predicted percentage of vote by .06% in a similar direction of the change.

(ii)**In terms of the parameters, state the null hypothesis that a 1% increase in A’s expenditures is offset by a 1% increase in B’s expenditures.** H0 : B1 = -B2

(iii) **Estimate the given model using the data in VOTE1.RAW and report the results 
in usual form. Do A’s expenditures affect the outcome? What about B’s expenditures? Can you use these results to test the hypothesis in part (ii)?** lexpendA has a coef of 6.08 and a std error of .382, which gives us a t statistic of 15.9 which means it is very significant both practically as statistically. Similarly, lexpendB has a coef of -6.61, a SD of .379 and a t of -17.463. It is also both practically and statistically significant and important. Both have similar results in the opposite direction, which is expected due to the non-zero sum aspect of an election. Further analysis is required to quantify and reject the hypothesis that a 1% increase in A is offset by 1% increase in B, such as applying a t test between the two parameters, which would require us to estimate a model that gives the t statistic for testing the hypothesis by reparametrization or to compute the standard error of B1-B2

(iv) Estimate a model that directly gives the t statistic for testing the hypothesis in part 
(ii). What do you conclude? (Use a two-sided alternative.)




voteA = β₀ + β₁lexpendA + β₂lexpendB + β₃prtystrA + u

Hypothesis: H0:B1=-B2

we sum and substract B2expend, to obtain the coefficient B1+B2 and prove if it is 0

voteA = β₀ + β₁lexpendA + β₂lexpendB + β₂lexpendA - β₂lexpendA + β₃prtystrA + u

voteA = β₀ + β₂(lexpendA + lexpendB) + (β₁ + β₂)lexpendA + β₃prtystrA + u

θ = β₂ (coeficiente de sum_expend)
γ = β₁ + β₂ (coeficiente de lexpendA)

voteA = β₀ + θ(lexpendA + lexpendB) + γ(lexpendA) + β₃prtystrA + u



In [23]:
# Answer to 4:
from scipy import stats  

# Crear nueva variable que representa la suma de los coeficientes
vote['sum_expend'] = vote['lexpendA'] + vote['lexpendB']

# Modelo reparametrizado
# voteA = β₀ + θ(lexpendA + lexpendB) + γ(lexpendA) + β₃prtystrA + u
# donde γ = (β₁ + β₂)/2 es el coeficiente que queremos probar si es 0
modelo_rep = ols('voteA ~ sum_expend + lexpendA + prtystrA', data=vote).fit()

print("Resultados del modelo reparametrizado:")
print("=====================================")
print(modelo_rep.summary())

# Calcular el p-valor para la prueba de dos colas
t_stat = modelo_rep.tvalues['lexpendA']
p_valor = 2 * (1 - stats.t.cdf(abs(t_stat), df=modelo_rep.df_resid))

print("\nPrueba de hipótesis H₀: β₁ = -β₂")
print("===================================")
print(f"t-estadístico: {t_stat:.4f}")
print(f"p-valor: {p_valor:.4e}")

Resultados del modelo reparametrizado:
                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Mon, 11 Nov 2024   Prob (F-statistic):           1.76e-57
Time:                        17:58:27   Log-Likelihood:                -596.86
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    

(iv) with a t statistic of 23.38 and a p value of almost 0, we can reject the null hipothesis that B1 = -B2. This means that the effects in expenditure aren't symmetric, where the expenditures by candidate B have a marginally bigger effect in voting.

## C2 
LSAT is the median LSAT score for the graduating class, GPA is the median college 
GPA for the class, libvol is the number of volumes in the law school library, cost is the annual cost of attending law school, and rank is a law school ranking (with rank 5 1 being 
the best).

In [29]:
import wooldridge as woo
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
from statsmodels.formula.api import ols
import seaborn as sns

data = woo.data('lawsch85')



In [30]:
modelo = ols('lsalary ~ LSAT + GPA + llibvol + lcost + rank', data=data)

resultados = modelo.fit()


print("\nResultados de la Regresión:")
print(resultados.summary())


Resultados de la Regresión:
                            OLS Regression Results                            
Dep. Variable:                lsalary   R-squared:                       0.842
Model:                            OLS   Adj. R-squared:                  0.836
Method:                 Least Squares   F-statistic:                     138.2
Date:                Mon, 11 Nov 2024   Prob (F-statistic):           2.93e-50
Time:                        18:28:35   Log-Likelihood:                 107.33
No. Observations:                 136   AIC:                            -202.7
Df Residuals:                     130   BIC:                            -185.2
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      8.3432  

(i)state and test the null hypothesis 
that the rank of law schools has no ceteris paribus effect on median starting salary.

H0: Brank = 0

the t statistic of rank is -9.541, which indicates that the rank of the school has a statistically significant effect on salary.The effect is low when compared to other variables. As it is a log-level relationship, each ranking point is equal to .33% in a change in the salary.

(ii) Are features of the incoming class of students—namely, LSAT and GPA—
individually or jointly significant for explaining salary? (Be sure to account for 
missing data on LSAT and GPA.)

Lsat doesn't seem to have individual significancy. To evaluate joint significance, one may use the F test to evaluate this.

In [33]:
import wooldridge as woo
import numpy as np
from statsmodels.formula.api import ols
import scipy.stats as stats

# Load and clean data
data = woo.data('lawsch85')
data_clean = data.dropna(subset=['LSAT', 'GPA', 'lsalary', 'llibvol', 'lcost', 'rank'])

# Estimate models
full_model = ols('lsalary ~ LSAT + GPA + llibvol + lcost + rank', data=data_clean).fit()
restricted_model = ols('lsalary ~ llibvol + lcost + rank', data=data_clean).fit()

# F-test for joint significance
n = full_model.nobs
k_full = full_model.df_model
k_restricted = restricted_model.df_model
q = k_full - k_restricted  # number of restrictions

F = ((full_model.rsquared - restricted_model.rsquared) / q) / \
    ((1 - full_model.rsquared) / (n - k_full - 1))
p_value = 1 - stats.f.cdf(F, q, n - k_full - 1)

# Print results
print(f"Sample size: {n}")
print(f"\nF-statistic: {F:.4f}")
print(f"p-value: {p_value:.4f}")

Sample size: 136.0

F-statistic: 9.9518
p-value: 0.0001


Both are jointly significant.

(iii) Test whether the size of the entering class (clsize) or the size of the faculty 
( faculty) needs to be added to this equation; carry out a single test. (Be careful to 
account for missing data on clsize and faculty.)

H0 : Bclsize, Bfaculty = 0
H1 : Bclsize, Bfaculty =/ 0



In [36]:
import wooldridge as woo
from statsmodels.formula.api import ols
import scipy.stats as stats

# Load and clean data
data = woo.data('lawsch85')
data_clean = data.dropna(subset=['LSAT', 'GPA', 'lsalary', 'llibvol', 'lcost', 'rank', 'clsize', 'faculty'])

# Extended model
extended_model = ols('lsalary ~ LSAT + GPA + llibvol + lcost + rank + clsize + faculty', 
                    data=data_clean).fit()

# Base model
base_model = ols('lsalary ~ LSAT + GPA + llibvol + lcost + rank', 
                data=data_clean).fit()

# F-test
n = base_model.nobs
k_extended = extended_model.df_model
k_base = base_model.df_model
q = k_extended - k_base

F = ((extended_model.rsquared - base_model.rsquared) / q) / \
    ((1 - extended_model.rsquared) / (n - k_extended - 1))
p_value = 1 - stats.f.cdf(F, q, n - k_extended - 1)

# Print results
print(f"Sample size: {n}")
print(f"\nF-test for joint significance of class size and faculty:")
print(f"F-statistic: {F:.4f}")
print(f"p-value: {p_value:.4f}")

print("\nCoefficients for added variables:")
print(f"{'Variable':<10} {'Coefficient':>12} {'Std Error':>12} {'t-stat':>10} {'p-value':>10}")
print("-" * 55)
print(f"{'clsize':<10} {extended_model.params['clsize']:12.4f} "
      f"{extended_model.bse['clsize']:12.4f} "
      f"{(extended_model.params['clsize']/extended_model.bse['clsize']):10.4f} "
      f"{extended_model.pvalues['clsize']:10.4f}")
print(f"{'faculty':<10} {extended_model.params['faculty']:12.4f} "
      f"{extended_model.bse['faculty']:12.4f} "
      f"{(extended_model.params['faculty']/extended_model.bse['faculty']):10.4f} "
      f"{extended_model.pvalues['faculty']:10.4f}")

Sample size: 131.0

F-test for joint significance of class size and faculty:
F-statistic: 0.9484
p-value: 0.3902

Coefficients for added variables:
Variable    Coefficient    Std Error     t-stat    p-value
-------------------------------------------------------
clsize           0.0001       0.0002     0.8741     0.3838
faculty          0.0001       0.0004     0.1687     0.8663


The F statistic is .9484 and the p-value is .3902
There is very little evidence that adding clsize and faculty will improve the model, so we fail to reject H0


(iv) What factors might influence the rank of the law school that are not included in the model?

- Bar passage rate
- Student acceptance ratio
- Number and quality of academic papers
- Ratio of employment after graduation

## C3

Use the log of the housing price as the 
dependent variable:
log(price) = b0 + b1sqrft + b2bdrms + u.
(i) You are interested in estimating and obtaining a confidence interval for the percentage change in price when a 150-square-foot bedroom is added to a house. In decimal 
form, this is θ = 150b1 + b2. Use the data in HPRICE1.RAW to estimate θ.

In [1]:
import wooldridge
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

from statsmodels.formula.api import ols

data = wooldridge.data('hprice1')

print (data.columns)

Index(['price', 'assess', 'bdrms', 'lotsize', 'sqrft', 'colonial', 'lprice',
       'lassess', 'llotsize', 'lsqrft'],
      dtype='object')


In [3]:

model = ols('lprice ~ sqrft + bdrms', data=data).fit()

beta1 = model.params['sqrft']
beta2 = model.params['bdrms']

# Theta 1 which represents a new bedroom of 150 sqrft
theta1 = 150 * beta1 + beta2

# Covariance matrix
vcov = model.cov_params()

# Standard error of theta1
var_theta1 = (150**2 * vcov.loc['sqrft', 'sqrft'] +
              vcov.loc['bdrms', 'bdrms'] +
              2 * 150 * vcov.loc['sqrft', 'bdrms'])
se_theta1 = np.sqrt(var_theta1)


In [4]:
# Confidence interval of price when theta1 (a 150 ft room) is added 
degrees_of_freedom = model.df_resid
t_critical = stats.t.ppf(0.975, degrees_of_freedom)
ci_lower = theta1 - t_critical * se_theta1
ci_upper = theta1 + t_critical * se_theta1

# Dataframe with the results

results_df = pd.DataFrame({
    'Métrica': ['θ₁ (efecto combinado)', 
                'Error Estándar', 
                'IC 95% - Límite Inferior',
                'IC 95% - Límite Superior',
                'Efecto Porcentual'],
    'Valor': [theta1, 
              se_theta1, 
              ci_lower, 
              ci_upper,
              (np.exp(theta1) - 1) * 100]  # Conversión a cambio porcentual exacto
})

print("\nResultados del Análisis:")
print("========================")
print(results_df.to_string(index=False))



Resultados del Análisis:
                 Métrica    Valor
   θ₁ (efecto combinado) 0.085801
          Error Estándar 0.026768
IC 95% - Límite Inferior 0.032580
IC 95% - Límite Superior 0.139022
       Efecto Porcentual 8.958985


In [7]:
# Imprimir estadísticas adicionales del modelo
print("\nEstadísticas del Modelo:")
print("=======================")
print(f"R² ajustado: {model.rsquared_adj:.4f}")
print(f"Estadístico F: {model.fvalue:.2f}")
print(f"Valor p (F): {model.f_pvalue:.4e}")



Estadísticas del Modelo:
R² ajustado: 0.5786
Estadístico F: 60.73
Valor p (F): 4.1679e-17


In [9]:

# Imprimir los coeficientes individuales y sus estadísticas
print("\nCoeficientes Individuales:")
print(model.summary().tables[1])


Coeficientes Individuales:
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.7660      0.097     49.112      0.000       4.573       4.959
sqrft          0.0004   4.32e-05      8.781      0.000       0.000       0.000
bdrms          0.0289      0.030      0.974      0.333      -0.030       0.088


# C4

This exercise asks to compute an F statistic to determine if parents' education are statistically significant on an incomplete dataset, so that we can compare the results of when it is computed on the data that is complete.

The unrestricted model is bwght = β₀ + β₁cigs + β₂parity + β₃faminc + β₄motheduc + β₅fatheduc + u

The sample size is 1388, of which only 1191 where used.  R² unrestricted: 0.0387. R² restricted: 0.0364. F: 1.42, mother and father education are statistically insignificant.

Next, we compute the same on the whole dataset, ignoring that there are missing values.

In [10]:
import wooldridge
import pandas as pd
import numpy as np
from scipy import stats

# Cargar los datos
data = wooldridge.data('bwght')

# Verificar el número total de observaciones
print(f"Número total de observaciones en el dataset: {len(data)}")

# 1. Modelo restringido usando todas las observaciones válidas para bwght, cigs, parity, faminc
# Primero veamos cuántas observaciones tenemos para cada variable
print("\nObservaciones disponibles por variable:")
for col in ['bwght', 'cigs', 'parity', 'faminc']:
    print(f"{col}: {data[col].count()} observaciones válidas")

# Ahora seleccionamos solo las variables necesarias para el modelo restringido
X_vars = ['cigs', 'parity', 'faminc']
mask = data[['bwght'] + X_vars].notna().all(axis=1)
model_restricted_full = data[mask]

X_restricted_full = model_restricted_full[X_vars]
y_full = model_restricted_full['bwght']

# Añadir constante
X_restricted_full = pd.concat([pd.Series(1, index=X_restricted_full.index, name='const'), 
                             X_restricted_full], axis=1)

# Calcular beta usando OLS
beta_restricted_full = np.linalg.inv(X_restricted_full.T @ X_restricted_full) @ (X_restricted_full.T @ y_full)

# Calcular valores ajustados y R² para modelo restringido con todas las observaciones válidas
y_hat_restricted_full = X_restricted_full @ beta_restricted_full
mean_y_full = np.mean(y_full)
r2_restricted_full = 1 - (np.sum((y_full - y_hat_restricted_full)**2) / 
                         np.sum((y_full - mean_y_full)**2))

# 2. Modelo restringido usando solo las 1,191 observaciones (del Example 4.9)
mask_complete = data[['bwght', 'cigs', 'parity', 'faminc', 'motheduc', 'fatheduc']].notna().all(axis=1)
model_complete = data[mask_complete]
X_restricted = model_complete[X_vars]
y = model_complete['bwght']

# Añadir constante
X_restricted = pd.concat([pd.Series(1, index=X_restricted.index, name='const'), 
                         X_restricted], axis=1)

# Calcular beta usando OLS
beta_restricted = np.linalg.inv(X_restricted.T @ X_restricted) @ (X_restricted.T @ y)

# Calcular valores ajustados y R² para modelo restringido
y_hat_restricted = X_restricted @ beta_restricted
mean_y = np.mean(y)
r2_restricted = 1 - (np.sum((y - y_hat_restricted)**2) / 
                     np.sum((y - mean_y)**2))

print("\nResultados:")
print(f"Observaciones usadas en Example 4.9 (con datos completos): {len(model_complete)}")
print(f"R² reportado en Example 4.9 (1,191 obs): 0.0364")
print(f"R² calculado con las mismas 1,191 obs: {r2_restricted:.4f}")
print(f"\nObservaciones usadas en modelo restringido (ejercicio C4): {len(model_restricted_full)}")
print(f"R² del modelo restringido (usando todas las obs válidas): {r2_restricted_full:.4f}")

Número total de observaciones en el dataset: 1388

Observaciones disponibles por variable:
bwght: 1388 observaciones válidas
cigs: 1388 observaciones válidas
parity: 1388 observaciones válidas
faminc: 1388 observaciones válidas

Resultados:
Observaciones usadas en Example 4.9 (con datos completos): 1191
R² reportado en Example 4.9 (1,191 obs): 0.0364
R² calculado con las mismas 1,191 obs: 0.0364

Observaciones usadas en modelo restringido (ejercicio C4): 1388
R² del modelo restringido (usando todas las obs válidas): 0.0348


R² is slightly lower when we use all the available samples. 

## C5


Starting from the following equation:
```log(salary) = 11.19 + 0.0689 years + 0.0126 gamesyr + 0.00098 bavg + 0.0144 hrunsyr + 0.0108 rbisyr```

We will analyze:
1. The effect of omitting rbisyr over the coefficient of hrunsyr
2. The impact of including runsyr, fldperc y sbasesyr to the model
3. The joint significance of bavg, fldperc y sbasesyr.

Dataset: MLB1.RAW, Wooldridge

In [11]:
import wooldridge as woo
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from scipy import stats

# Cargar los datos
mlb = woo.data('mlb1')

print(mlb.columns)


Index(['salary', 'teamsal', 'nl', 'years', 'games', 'atbats', 'runs', 'hits',
       'doubles', 'triples', 'hruns', 'rbis', 'bavg', 'bb', 'so', 'sbases',
       'fldperc', 'frstbase', 'scndbase', 'shrtstop', 'thrdbase', 'outfield',
       'catcher', 'yrsallst', 'hispan', 'black', 'whitepop', 'blackpop',
       'hisppop', 'pcinc', 'gamesyr', 'hrunsyr', 'atbatsyr', 'allstar',
       'slugavg', 'rbisyr', 'sbasesyr', 'runsyr', 'percwhte', 'percblck',
       'perchisp', 'blckpb', 'hispph', 'whtepw', 'blckph', 'hisppb',
       'lsalary'],
      dtype='object')


In [12]:
# Original model
X_orig = sm.add_constant(mlb[['years', 'gamesyr', 'bavg', 'hrunsyr', 'rbisyr']])
y = mlb['lsalary']

modelo_original = sm.OLS(y, X_orig).fit()

In [13]:
# (i) model without rbisyr
X_sin_rbi = sm.add_constant(mlb[['years', 'gamesyr', 'bavg', 'hrunsyr']])
modelo_sin_rbi = sm.OLS(y, X_sin_rbi).fit()

In [14]:
# Model with additional vars
X_completo = sm.add_constant(mlb[['years', 'gamesyr', 'bavg', 'hrunsyr', 'runsyr', 'fldperc', 'sbasesyr']])

modelo_completo = sm.OLS(y, X_completo).fit()

In [15]:
def print_results_comparison():
    print("=== Comparación de Modelos ===\n")
    
    # Resultados del modelo original
    print("Modelo Original (ecuación 4.31):")
    print(modelo_original.summary().tables[1])
    print("\nR-cuadrado:", round(modelo_original.rsquared, 4))
    print("\n" + "="*50 + "\n")
    
    # Resultados del modelo sin rbisyr
    print("Modelo sin rbisyr:")
    print(modelo_sin_rbi.summary().tables[1])
    print("\nR-cuadrado:", round(modelo_sin_rbi.rsquared, 4))
    print("\n" + "="*50 + "\n")
    
    # Resultados del modelo completo
    print("Modelo Completo (con variables adicionales):")
    print(modelo_completo.summary().tables[1])
    print("\nR-cuadrado:", round(modelo_completo.rsquared, 4))

In [16]:
def test_joint_significance():
    # Prueba F para la significancia conjunta de bavg, fldperc y sbasesyr
    # Crear modelos restringido y no restringido
    variables_restriccion = ['bavg', 'fldperc', 'sbasesyr']
    variables_base = ['years', 'gamesyr', 'hrunsyr', 'runsyr']
    
    # Modelo no restringido (completo)
    X_no_restringido = sm.add_constant(mlb[variables_base + variables_restriccion])
    modelo_no_restringido = sm.OLS(y, X_no_restringido).fit()
    
    # Modelo restringido (sin las variables a probar)
    X_restringido = sm.add_constant(mlb[variables_base])
    modelo_restringido = sm.OLS(y, X_restringido).fit()
    
    # Calcular estadístico F
    n = len(y)
    k_no_restringido = len(X_no_restringido.columns)
    k_restringido = len(X_restringido.columns)
    q = k_no_restringido - k_restringido
    
    F_stat = ((modelo_restringido.ssr - modelo_no_restringido.ssr) / q) / \
             (modelo_no_restringido.ssr / (n - k_no_restringido))
    
    p_value = 1 - stats.f.cdf(F_stat, q, n - k_no_restringido)
    
    print("\n=== Prueba F de Significancia Conjunta ===")
    print(f"Variables probadas: {', '.join(variables_restriccion)}")
    print(f"Estadístico F: {F_stat:.4f}")
    print(f"Valor p: {p_value:.4f}")
    print(f"Grados de libertad: ({q}, {n - k_no_restringido})")

In [17]:
print_results_comparison()
test_joint_significance()

=== Comparación de Modelos ===

Modelo Original (ecuación 4.31):
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.1924      0.289     38.752      0.000      10.624      11.760
years          0.0689      0.012      5.684      0.000       0.045       0.093
gamesyr        0.0126      0.003      4.742      0.000       0.007       0.018
bavg           0.0010      0.001      0.887      0.376      -0.001       0.003
hrunsyr        0.0144      0.016      0.899      0.369      -0.017       0.046
rbisyr         0.0108      0.007      1.500      0.134      -0.003       0.025

R-cuadrado: 0.6278


Modelo sin rbisyr:
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.0209      0.266     41.476      0.000      10.498      11.544
years          0.0677    

Hrunsyr increased its t value from 0.899 to 4.964, becoming highly significant. This is likely due to a multicolineality problem, which is logic as home runs will always result in at least 1 Run Batted in.

Since Runs batted in depend so much on other variables, and R² barely changed, one can conclude that rbisyr and hrunsyr capture similar information. 

runsyr, fldperc and sbasesyr have the following coefficients:0.0174, 0.0010 and -0.0064, with respective t statistics of 3.43, 0.516 and -1.238. Runsyr seems to be significant, but one might have doubts about bavg, fldperc and sbasesyr. The joint significance of  bavg, fldperc and sbasesyr variables are very low, with an f statistic of 0.685 and a p value of .5617, which means that there is not enough evidence that they are providing jointly any information to the model, and should be removed.

## C6

Using the data from WAGE2, estimate a log-level equation of salaries. Prove if general experience has the same effect as an extra year of remaining with the present employer. 

h0: β₂ = β₃
H₁: β₂ ≠ β₃

Significance: 5%
Confidence interval: 95%

In [18]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from wooldridge import data
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Cargar los datos
df = data('wage2')

In [20]:
X = df[['educ', 'exper', 'tenure']]
X = sm.add_constant(X)
y = df['lwage']

modelo = sm.OLS(y, X).fit()

In [24]:
# Imprimir resultados
print("\nResultados del Modelo de Regresión:")
print("===================================")
print(modelo.summary())
print("\nFactores de Inflación de Varianza (VIF):")
print("=========================================")

# Pruebas adicionales de diagnóstico
print("\nPruebas de Diagnóstico:")
print("======================")
print(f"R-cuadrado: {modelo.rsquared:.4f}")
print(f"R-cuadrado ajustado: {modelo.rsquared_adj:.4f}")
print(f"Estadístico F: {modelo.fvalue:.4f}")
print(f"Prob (F-statistic): {modelo.f_pvalue:.4f}")

# Interpretación de los coeficientes en términos porcentuales
print("\nInterpretación de los Coeficientes:")
print("==================================")
for var, coef in zip(X.columns, modelo.params):
    if var != 'const':
        print(f"\n{var}: Un aumento de 1 año en {var} está asociado con un cambio de {(np.exp(coef)-1)*100:.2f}% en el salario, manteniendo las otras variables constantes")


Resultados del Modelo de Regresión:
                            OLS Regression Results                            
Dep. Variable:                  lwage   R-squared:                       0.155
Model:                            OLS   Adj. R-squared:                  0.152
Method:                 Least Squares   F-statistic:                     56.97
Date:                Tue, 12 Nov 2024   Prob (F-statistic):           8.12e-34
Time:                        17:48:47   Log-Likelihood:                -438.84
No. Observations:                 935   AIC:                             885.7
Df Residuals:                     931   BIC:                             905.0
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          

In [25]:
beta_exper = modelo.params['exper']
beta_tenure = modelo.params['tenure']
cov_matrix = modelo.cov_params()

# Calcular la varianza de la diferencia
# Var(β₂ - β₃) = Var(β₂) + Var(β₃) - 2Cov(β₂,β₃)

var_diff = (cov_matrix.loc['exper', 'exper'] +
            cov_matrix.loc['tenure', 'tenure'] -
            2 * cov_matrix.loc['exper', 'tenure'])

# Calcular estadístico t
diff = beta_exper - beta_tenure
se_diff = np.sqrt(var_diff)
t_stat = diff / se_diff

# Grados de libertad
df_reg = modelo.df_resid

# Valor crítico
t_crit = stats.t.ppf(0.975, df_reg)
p_value = 2 * (1 - stats.t.cdf(abs(t_stat), df_reg))

# Intervalo de confianza
ic_lower = diff - t_crit * se_diff
ic_upper = diff + t_crit * se_diff

In [26]:
# Imprimir resultados
print("Prueba de Hipótesis: β_exper = β_tenure")
print("=======================================")
print(f"\nCoeficiente de experiencia (β_exper): {beta_exper:.4f}")
print(f"Coeficiente de tenure (β_tenure): {beta_tenure:.4f}")
print(f"Diferencia (β_exper - β_tenure): {diff:.4f}")
print(f"\nError estándar de la diferencia: {se_diff:.4f}")
print(f"Estadístico t: {t_stat:.4f}")
print(f"Grados de libertad: {df_reg}")
print(f"Valor crítico (t_{df_reg},0.025): ±{t_crit:.4f}")
print(f"P-valor: {p_value:.4f}")
print("\nIntervalo de Confianza 95%")
print("==========================")
print(f"Límite inferior: {ic_lower:.4f}")
print(f"Límite superior: {ic_upper:.4f}")

Prueba de Hipótesis: β_exper = β_tenure

Coeficiente de experiencia (β_exper): 0.0153
Coeficiente de tenure (β_tenure): 0.0134
Diferencia (β_exper - β_tenure): 0.0020

Error estándar de la diferencia: 0.0047
Estadístico t: 0.4119
Grados de libertad: 931.0
Valor crítico (t_931.0,0.025): ±1.9625
P-valor: 0.6805

Intervalo de Confianza 95%
Límite inferior: -0.0074
Límite superior: 0.0113


With a t value of .4119 there is not enough evidence that experience is different than tenure, so we can conclude that Bexper = Btenure

## C7

In [31]:
import wooldridge as woo
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns

# Cargar los datos
data = woo.data('twoyear')
print(data.columns)


Index(['female', 'phsrank', 'BA', 'AA', 'black', 'hispanic', 'id', 'exper',
       'jc', 'univ', 'lwage', 'stotal', 'smcity', 'medcity', 'submed',
       'lgcity', 'sublg', 'vlgcity', 'subvlg', 'ne', 'nc', 'south', 'totcoll'],
      dtype='object')


In [33]:
# Calcular estadísticas descriptivas para phsrank
phsrank_stats = {
    'Mínimo': data['phsrank'].min(),
    'Máximo': data['phsrank'].max(),
    'Promedio': data['phsrank'].mean(),
    'Mediana': data['phsrank'].median(),
    'Desviación estándar': data['phsrank'].std()
}

stats_df = pd.DataFrame.from_dict(phsrank_stats, orient='index', columns=['Valor'])
print("Estadísticas descriptivas para phsrank (percentil de rango en la escuela secundaria):")
print(stats_df)


Estadísticas descriptivas para phsrank (percentil de rango en la escuela secundaria):
                         Valor
Mínimo                0.000000
Máximo               99.000000
Promedio             56.157031
Mediana              50.000000
Desviación estándar  24.272964


In [35]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

# log(wage) = β₀ + β₁jc + β₂univ + β₃exper + β₄phsrank + u
model = ols('lwage ~ jc + univ + exper + phsrank', data=data)
results = model.fit()
print(results.summary().tables[1])

# Calcular R-cuadrado
print("\nR-cuadrado:", round(results.rsquared, 4))
print("Número de observaciones:", results.nobs)



                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.4587      0.024     61.756      0.000       1.412       1.505
jc             0.0662      0.007      9.671      0.000       0.053       0.080
univ           0.0755      0.003     29.496      0.000       0.070       0.080
exper          0.0049      0.000     31.360      0.000       0.005       0.005
phsrank        0.0003      0.000      1.269      0.204      -0.000       0.001

R-cuadrado: 0.2226
Número de observaciones: 6763.0


phsrank:

Not statistically significant (t = 1.26)
Increasing 10 percentile points produce a change of 0.3% in salary (.0003 × 10 × 100)

Adding phsrank doesn't change the conclusions of the model without it.

## C8

In [38]:
import pandas as pd
import numpy as np
import wooldridge as wd
from statsmodels.formula.api import ols
import statsmodels.api as sm

# Cargar los datos
data = wd.data('401ksubs')

# Filtrar solo hogares unipersonales
single_household = data[data['fsize'] == 1].copy()

# (i) Número de hogares unipersonales
n_singles = len(single_household)
print(f"Número de hogares unipersonales: {n_singles}")

# (ii) Estimación del modelo OLS
# Crear el modelo
model = ols('nettfa ~ inc + age', data=single_household)
results = model.fit()

# Imprimir resultados
print("\nResultados de la Regresión:")
print(results.summary().tables[1])

# Calcular R-cuadrado
r2 = results.rsquared
r2_adj = results.rsquared_adj

print(f"\nR-cuadrado: {r2:.4f}")
print(f"R-cuadrado ajustado: {r2_adj:.4f}")

# Diagnósticos básicos
print("\nDiagnósticos del modelo:")
print(f"Test F (p-valor): {results.f_pvalue:.4f}")
print(f"Estadístico F: {results.fvalue:.4f}")

Número de hogares unipersonales: 2017

Resultados de la Regresión:
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -43.0398      4.080    -10.548      0.000     -51.042     -35.038
inc            0.7993      0.060     13.382      0.000       0.682       0.916
age            0.8427      0.092      9.158      0.000       0.662       1.023

R-cuadrado: 0.1193
R-cuadrado ajustado: 0.1185

Diagnósticos del modelo:
Test F (p-valor): 0.0000
Estadístico F: 136.4648


In [39]:
# Extraer coeficiente y error estándar para age
beta_2 = results.params['age']
se_beta_2 = results.bse['age']

# Calcular el estadístico t para H0: β₂ = 1
t_stat = (beta_2 - 1) / se_beta_2

# Calcular el p-valor (prueba de dos colas)
p_value = 2 * (1 - stats.t.cdf(abs(t_stat), df=len(single_household)-3))

print(f"Coeficiente de age (β₂): {beta_2:.4f}")
print(f"Error estándar: {se_beta_2:.4f}")
print(f"Estadístico t: {t_stat:.4f}")
print(f"P-valor: {p_value:.4f}")

# Comparación con regresión simple
simple_model = ols('nettfa ~ inc', data=single_household)
simple_results = simple_model.fit()

print("\nComparación de coeficientes de ingreso:")
print(f"Modelo múltiple: {results.params['inc']:.4f}")
print(f"Modelo simple: {simple_results.params['inc']:.4f}")

Coeficiente de age (β₂): 0.8427
Error estándar: 0.0920
Estadístico t: -1.7099
P-valor: 0.0874

Comparación de coeficientes de ingreso:
Modelo múltiple: 0.7993
Modelo simple: 0.8207


The intercept has no practical meaning in this problem. 

There is no evidence to reject that the effect on age is different than 1.

The coefficient has little change over a simple regression model, which suggests low correlation between these two variables

## C9

This problem investigates if racial discrimination exists in soft drink prices at fast-food restaurants by analyzing whether the percentage of black population in a zip code significantly influences prices, while controlling for socioeconomic variables such as income, poverty rates, housing values.


In [41]:
# Importamos las bibliotecas necesarias
import wooldridge as woo
import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy import stats

# Cargamos los datos
data = woo.data('discrim')
print (data.columns)

Index(['psoda', 'pfries', 'pentree', 'wagest', 'nmgrs', 'nregs', 'hrsopen',
       'emp', 'psoda2', 'pfries2', 'pentree2', 'wagest2', 'nmgrs2', 'nregs2',
       'hrsopen2', 'emp2', 'compown', 'chain', 'density', 'crmrte', 'state',
       'prpblck', 'prppov', 'prpncar', 'hseval', 'nstores', 'income', 'county',
       'lpsoda', 'lpfries', 'lhseval', 'lincome', 'ldensity', 'NJ', 'BK',
       'KFC', 'RR'],
      dtype='object')


In [44]:
# Verificamos valores faltantes o infinitos
print("Información sobre valores faltantes antes de la limpieza:")
print(data[['lpsoda', 'prpblck', 'lincome', 'prppov']].isna().sum())
print("\nInformación sobre valores infinitos antes de la limpieza:")
print(np.isinf(data[['lpsoda', 'prpblck', 'lincome', 'prppov']]).sum())


Información sobre valores faltantes antes de la limpieza:
lpsoda     8
prpblck    1
lincome    1
prppov     1
dtype: int64

Información sobre valores infinitos antes de la limpieza:
lpsoda     0
prpblck    0
lincome    0
prppov     0
dtype: int64


In [46]:
variables_modelo = ['lpsoda', 'prpblck', 'lincome', 'prppov']
data_clean = data[variables_modelo].replace([np.inf, -np.inf], np.nan).dropna()
print("\nTamaño del dataset:")
print(f"Original: {len(data)}")
print(f"Después de limpiar: {len(data_clean)}")



Tamaño del dataset:
Original: 410
Después de limpiar: 401


In [47]:

X = data_clean[['prpblck', 'lincome', 'prppov']]
X = sm.add_constant(X)  # Añadimos la constante
y = data_clean['lpsoda']

# Estimamos el modelo
modelo = sm.OLS(y, X).fit()

# Imprimimos los resultados
print("Resultados de la Regresión:")
print("==========================")
print(modelo.summary())

Resultados de la Regresión:
                            OLS Regression Results                            
Dep. Variable:                 lpsoda   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.080
Method:                 Least Squares   F-statistic:                     12.60
Date:                Tue, 12 Nov 2024   Prob (F-statistic):           6.92e-08
Time:                        18:52:00   Log-Likelihood:                 439.04
No. Observations:                 401   AIC:                            -870.1
Df Residuals:                     397   BIC:                            -854.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.4633   

In [48]:
# Análisis específico del coeficiente de prpblck (β1)
beta1 = modelo.params['prpblck']
se_beta1 = modelo.bse['prpblck']
t_stat = beta1 / se_beta1
p_value = 2 * (1 - stats.t.cdf(abs(t_stat), df=modelo.df_resid))

In [49]:
print("\nAnálisis detallado del coeficiente de prpblck (β1):")
print("=================================================")
print(f"Coeficiente (β1): {beta1:.4f}")
print(f"Error estándar: {se_beta1:.4f}")
print(f"Estadístico t: {t_stat:.4f}")
print(f"Valor p: {p_value:.4f}")


Análisis detallado del coeficiente de prpblck (β1):
Coeficiente (β1): 0.0728
Error estándar: 0.0307
Estadístico t: 2.3735
Valor p: 0.0181


In [50]:
# Análisis de significancia
alpha_5 = 0.05
alpha_1 = 0.01

print("\nPruebas de Significancia:")
print("========================")
print(f"Valor crítico al 5%: {stats.t.ppf(1-alpha_5/2, df=modelo.df_resid):.4f}")
print(f"Valor crítico al 1%: {stats.t.ppf(1-alpha_1/2, df=modelo.df_resid):.4f}")
print(f"Significativo al 5%: {'Sí' if p_value < alpha_5 else 'No'}")
print(f"Significativo al 1%: {'Sí' if p_value < alpha_1 else 'No'}")



Pruebas de Significancia:
Valor crítico al 5%: 1.9660
Valor crítico al 1%: 2.5883
Significativo al 5%: Sí
Significativo al 1%: No


In [51]:
# Intervalos de confianza
ci_95 = modelo.conf_int().loc['prpblck']
ci_99 = modelo.conf_int(alpha=0.01).loc['prpblck']

print("\nIntervalos de Confianza para β1:")
print("===============================")
print(f"Intervalo de confianza al 95%: [{ci_95[0]:.4f}, {ci_95[1]:.4f}]")
print(f"Intervalo de confianza al 99%: [{ci_99[0]:.4f}, {ci_99[1]:.4f}]")


Intervalos de Confianza para β1:
Intervalo de confianza al 95%: [0.0125, 0.1331]
Intervalo de confianza al 99%: [-0.0066, 0.1522]


In [52]:
# Calculamos la correlación entre lincome y prppov
correlation = data_clean['lincome'].corr(data_clean['prppov'])

print("Correlación entre log(income) y prppov:")
print(f"r = {correlation:.4f}")

Correlación entre log(income) y prppov:
r = -0.8402


In [55]:

data_clean = data[['lpsoda', 'prpblck', 'lincome', 'prppov', 'lhseval']].replace([np.inf, -np.inf], np.nan).dropna()

# Añadimos log(hseval)
X_new = data_clean[['prpblck', 'lincome', 'prppov', 'lhseval']]
X_new = sm.add_constant(X_new)
y = data_clean['lpsoda']

modelo_nuevo = sm.OLS(y, X_new).fit()

print("\nResultados del nuevo modelo (incluyendo lhseval):")
print("===============================================")
print(modelo_nuevo.summary())


Resultados del nuevo modelo (incluyendo lhseval):
                            OLS Regression Results                            
Dep. Variable:                 lpsoda   R-squared:                       0.184
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     22.31
Date:                Tue, 12 Nov 2024   Prob (F-statistic):           1.24e-16
Time:                        18:58:13   Log-Likelihood:                 461.55
No. Observations:                 401   AIC:                            -913.1
Df Residuals:                     396   BIC:                            -893.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c

In [56]:
# Parte (iv) - Prueba F para significancia conjunta de lincome y prppov
# Creamos el modelo restringido (sin lincome y prppov)
X_restricted = data_clean[['prpblck', 'lhseval']]
X_restricted = sm.add_constant(X_restricted)
modelo_restringido = sm.OLS(y, X_restricted).fit()

# Realizamos la prueba F
f_stat = ((modelo_restringido.ssr - modelo_nuevo.ssr) / 2) / (modelo_nuevo.ssr / modelo_nuevo.df_resid)
p_value_f = 1 - stats.f.cdf(f_stat, 2, modelo_nuevo.df_resid)

print("\nPrueba de significancia conjunta para lincome y prppov:")
print(f"Estadístico F: {f_stat:.4f}")
print(f"Valor p: {p_value_f:.4f}")


Prueba de significancia conjunta para lincome y prppov:
Estadístico F: 3.5227
Valor p: 0.0304


The analysis reveals that the prpblck coefficient is significant at 5% but not at 1% in the initial model.  (β₁=0.0728, p=0.0181)

There is a strong negative correlation between lincome and prppov, (r=-0.8402) indicating multicollinearity. 

adding lhseval (avg housing prices in the zip code), the prpblck coefficient becomes more significant  (β₁=0.0976, p=0.001) while lincome and prppov lose individual significance, but mantain joint significance  (p=0.0304)

The first model, which only has prpblck (p=0.018) might be capturing both racial and socioeconomic effects.

The second model, with lhseval, should be capturing better the socioeconomic effect. Including this to the model allows us to separate both racial and economic discrimination. Even though there is a clear multicolinallity problem, which might require more discussion to select the proper model.

## C10

The study examines the relationship between salaries and benefits in the school system, specifically analyzing whether there is a trade-off between these compensation components. It seeks to determine if schools offering higher benefits (measured as a benefits-to-salary ratio) tend to pay lower salaries, controlling for factors such as school size (enrollment), staff, and students' socioeconomic status (lunch).

In [57]:
import wooldridge as woo
import pandas as pd
import numpy as np
from scipy import stats
import statsmodels.api as sm

# Cargar los datos
data = woo.data('elem94_95')

# (i) Regresión simple de lavgsal sobre bs
X = data['bs']
y = data['lavgsal']
X = sm.add_constant(X)
model1 = sm.OLS(y, X).fit()
print("Modelo 1: Regresión Simple")
print(model1.summary())

# Test de hipótesis para β = -1
beta_hat = model1.params[1]
se_beta = model1.bse[1]
t_stat = (beta_hat - (-1)) / se_beta
p_value = 2 * (1 - stats.t.cdf(abs(t_stat), model1.df_resid))

print("\nTest de hipótesis para β = -1:")
print(f"t-estadístico: {t_stat:.4f}")
print(f"p-valor: {p_value:.4f}")

# (ii) Regresión múltiple añadiendo lenrol y lstaff
X2 = data[['bs', 'lenrol', 'lstaff']]
X2 = sm.add_constant(X2)
model2 = sm.OLS(y, X2).fit()
print("\nModelo 2: Regresión Múltiple")
print(model2.summary())

# (v) Añadir la variable lunch
X3 = data[['bs', 'lenrol', 'lstaff', 'lunch']]
X3 = sm.add_constant(X3)
model3 = sm.OLS(y, X3).fit()
print("\nModelo 3: Regresión con lunch")
print(model3.summary())

# Calcular R² para cada modelo
r2_1 = model1.rsquared
r2_2 = model2.rsquared
r2_3 = model3.rsquared

print("\nComparación de R²:")
print(f"R² Modelo 1: {r2_1:.4f}")
print(f"R² Modelo 2: {r2_2:.4f}")
print(f"R² Modelo 3: {r2_3:.4f}")

Modelo 1: Regresión Simple
                            OLS Regression Results                            
Dep. Variable:                lavgsal   R-squared:                       0.015
Model:                            OLS   Adj. R-squared:                  0.015
Method:                 Least Squares   F-statistic:                     28.23
Date:                Tue, 12 Nov 2024   Prob (F-statistic):           1.21e-07
Time:                        19:18:51   Log-Likelihood:                 85.171
No. Observations:                1848   AIC:                            -166.3
Df Residuals:                    1846   BIC:                            -155.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.7479    

C:\Users\Gerardo\AppData\Local\Temp\ipykernel_4760\3571619949.py:19: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  beta_hat = model1.params[1]
C:\Users\Gerardo\AppData\Local\Temp\ipykernel_4760\3571619949.py:20: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  se_beta = model1.bse[1]


The analyses confirm a negative trade-off between salaries and benefits: for each unit increase in the benefits-to-salary ratio, salaries decrease by approximately 51.6% (in the full model). This relationship remains significant even after controlling for school size, staff, and student socioeconomic characteristics, although its magnitude reduces when including these controls (from -79.5% in the simple model to -51.6% in the full model). The final model explains 48.8% of the variation in salaries (R² = 0.4882).

(i) We can't reject that β = -1

(ii) b/s becomes less negative when we add other variables. It stays significant (p-value < .001) and R² increases substantially

(iii) when adding the variables, the standard error reduces. When compared to multicolineality, the reduction in variance is stronger. This can be explained by the drastic increase in R²

(iv) lstaff seems to be negative and large in magnitude. This reflects a tradeoff between the number of workers and the salary

(v) adding lunch decreases salaries because lunch program is a proxy to lower socioeconomic power. Therefore, if a school is on a zone with higher poverty levels, then the salaries are expected to drop.



## C11

This study analyzes the determinants of the years of education in a person, using parental education, student's ability (including a cuadratic term to capture non-linear effects) and college tuition costs as explanatory variables.

In [60]:
# Importamos las bibliotecas necesarias
import wooldridge as woo
import pandas as pd
import numpy as np
from scipy import stats
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns

# Cargamos los datos
data = woo.data('htv')

# (i) Estimamos el modelo de regresión
# Creamos la variable abil al cuadrado
data['abil2'] = data['abil']**2

# Creamos el modelo con statsmodels
X = sm.add_constant(data[['motheduc', 'fatheduc', 'abil', 'abil2']])
y = data['educ']

# Estimamos el modelo
modelo1 = sm.OLS(y, X).fit()
print("Resultados del modelo inicial:")
print(modelo1.summary())

# Test para relación lineal vs cuadrática (H0: coef_abil2 = 0)
print("\nTest para relación cuadrática:")
print(f"t-estadístico para abil2: {modelo1.tvalues['abil2']:.4f}")
print(f"p-valor para abil2: {modelo1.pvalues['abil2']:.4f}")

# (ii) Test H0: β1 = β2
# Creamos una matriz R para el test de restricción lineal
R = np.array([[0, 1, -1, 0, 0]])  # motheduc - fatheduc = 0
test_result = modelo1.t_test(R)
print("\nTest de igualdad entre coeficientes de motheduc y fatheduc:")
print(f"t-estadístico: {float(test_result.statistic):.4f}")
print(f"p-valor: {float(test_result.pvalue):.4f}")

# (iii) Agregamos variables de matrícula
X2 = sm.add_constant(data[['motheduc', 'fatheduc', 'abil', 'abil2', 'tuit17', 'tuit18']])
modelo2 = sm.OLS(y, X2).fit()

# Test conjunto para las variables de matrícula
R2 = np.zeros((2, 7))  # 7 es el número de parámetros en el modelo2
R2[0, 5] = 1  # coeficiente de tuit17
R2[1, 6] = 1  # coeficiente de tuit18
test_result2 = modelo2.f_test(R2)
print("\nTest conjunto para variables de matrícula:")
print(f"F-estadístico: {float(test_result2.statistic):.4f}")
print(f"p-valor: {float(test_result2.pvalue):.4f}")

# (iv) Correlación entre tuit17 y tuit18
corr = data['tuit17'].corr(data['tuit18'])
print(f"\nCorrelación entre tuit17 y tuit18: {corr:.4f}")

# Creamos el promedio de matrícula
data['tuit_avg'] = (data['tuit17'] + data['tuit18']) / 2

# Modelo con promedio de matrícula
X3 = sm.add_constant(data[['motheduc', 'fatheduc', 'abil', 'abil2', 'tuit_avg']])
modelo3 = sm.OLS(y, X3).fit()
print("\nModelo con promedio de matrícula:")
print(modelo3.summary())



Resultados del modelo inicial:
                            OLS Regression Results                            
Dep. Variable:                   educ   R-squared:                       0.444
Model:                            OLS   Adj. R-squared:                  0.443
Method:                 Least Squares   F-statistic:                     244.9
Date:                Tue, 12 Nov 2024   Prob (F-statistic):          1.34e-154
Time:                        19:42:26   Log-Likelihood:                -2436.6
No. Observations:                1230   AIC:                             4883.
Df Residuals:                    1225   BIC:                             4909.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          8.2402

C:\Users\Gerardo\AppData\Local\Temp\ipykernel_4760\2605916308.py:36: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  print(f"t-estadístico: {float(test_result.statistic):.4f}")


The model explains 44.4% of the variation in years of education, where both parental education and student's ability are statistically significant. There is a stronger effect in mother's education(0.19 years) compared to father's education (0.11 years). The relationship between ability and education is quadratic, confirmed by a significant abil^2 term.  Tuition costs, however, show no significant effect on years of education, even when using the average of both periods due to their high correlation.

(i) H₀: β₄abil² = 0 H₁: β₄ ≠ 0. With a t-statistic of 6.093, we know the relation is quadratic.
(ii) H0: β1 = β2. With a t-statistic of 1.93, we cant reject that these coefficients (mother and father educ) are equal. However, we have evidence to reject it at 10%
(iii) We cant reject that tuition isn't 0.
(iv) tuit17 and tuit18 are highly correlated. It is better to use the average of both due to their correlation. Averaging both years (which are highly correlated at 0.9808) is preferred as it reduces measurement error, provides a more representative measure of the true tuition costs faced by students, and still maintains model parsimony while avoiding multicollinearity issues.
(v) Including the average tuition didn't change R², and isn't statistically significant.

